### DISCUSSION TITLE
Possible bug

### DISCUSSION THREAD

**Discussion by BernhardGlueck** [ravendb/ravendb#18971] [General]
In my code i have two document types defined like so: ( abbreviated ) 

```

public class TreeContentCollectionDocument
{

    public required string Id { get; init; }
    
    public required string RootNodeId { get; init; }
}


public sealed class TreeContentCollectionNodeDocument
{

    public required string Id { get; init; }
    public required List<string> ChildNodeIds { get; init; }

}
```

I index the treecontentcollection Documents as part of a multi map reduce index. 
Here is the relevant part of that index: 

```

 AddMap<TreeContentCollectionDocument>(documents =>
                                                  from collection in documents
                                                  let rootNode = LoadDocument<TreeContentCollectionNodeDocument>(collection.RootNodeId)
                                                  from node in Recurse(rootNode,t => t.ChildNodeIds)
                                                               .Select(c => LoadDocument<TreeContentCollectionNodeDocument>(c))
                                                  from entry in node.Entries
                                                  select new
                                                  {
                                                      ContentId = entry.Id,
                                                      RefCount = 1,
                                                      LastReferencedAt = entry.CreatedAt,
                                                  });

```

Now i sometimes get indexing errors on this index. Here is the stack trace ( abbrev ) 

```
Failed to execute mapping function on TreeContentCollections/1A743BDC3961B9F96C5FE2A6FFE151266229FCF12063EC0956858B105B231FB5. Exception: System.InvalidOperationException: LoadDocument may only be called with a string or an enumerable, but was called with a parameter of type System.Collections.Generic.KeyValuePair`2[[System.Object, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Object, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]: [@metadata, {"@collection":"TreeContentCollectionNodeDocuments","Raven-Clr-Type":"DataPot.Server.Providers.Raven.Vcs.Content.Collections.TreeContentCollectionNodeDocument, DataPot.Server"}]
```

I have debugged this is a far as i can see, and it seems that when the "ChildNodeId" list is empty
```
ChildNodeIds: [] 
```

on the server side than the Recurse or Select call that walks them deduces the type of them to be KeyValuePair<object,object> instead of string ... I get that this is qutie a tricky case, since the type information would need to come from the index definition and not the actual json document ..... ( i think you translate indices to javascript since 4.0 ? ) 

I have also tried adding an explicit cast to string which did not help either ....

**Comment by grisha-kotler** [comment]
Can you share the index definition that you have on the server?

**Reply by BernhardGlueck** [reply]
I could send you a dump of the database including the documents and the index definitions as well as a debug package... 
But i would like not to do that in public ...  its only about 500 kb in total. The database dump exhibits the mentioned problem.

**Reply by arekpalinski** [reply]
Hi @BernhardGlueck,

You can send it to support@ravendb.net

**Comment by Lwiel** [comment]
The issue is that in `Recurse(rootNode, t => t.ChildNodeIds)` the `rootNode` is of `TreeContentCollectionNodeDocument` type, but provided lambda returns a list of strings. This means that in second call it's trying to access `ChildNodeIds` property from a list of strings, and the exception is thrown.

What exactly do you want your map function to do? Note that recursively loading documents in the indexing function can be problematic, because if you change a single document you'll have to re-index a huge number of related documents. You can read more about this [in our documentation](https://ravendb.net/docs/article-page/6.0/csharp/studio/database/indexes/indexing-performance#loaddocument-misuse).

**Comment by BernhardGlueck** [comment]
Thank you for your help ! Yes that indeed was the issue, funny however that it only failed in about 0.01 % of indexing attempts and worked correctly the other times ? 

That indexing these tree structures and their child nodes is a fanout at that modifying just one document would trigger a reindex is clear, however our tree nodes are immutable including their child relationships ( think merkle tree style hashing ) thus once indexed the whole tree will never change.
